### DISCUSSION TITLE
Binding to attached property chain

### DISCUSSION THREAD

**Discussion by kazo0** [unoplatform/uno#11656] [General]
[From Discord:](https://discord.com/channels/372137812037730304/550416151172087808/1083867397267652690)

> Friends, I am trying Uno for WinUI, an the following binding is not working, (even though it is correct)
Text="{Binding Path=(local:AttachedProps.MyElement).(local:AttachedProps.MyStr), RelativeSource={RelativeSource Mode=Self}}"
Basically, i have an attached property on an element pointed by an attached property on the TextBlock

> Here is the code:

```xml
    <StackPanel Orientation="Vertical"
                HorizontalAlignment="Center"
                VerticalAlignment="Center">
        <TextBlock x:Name="TextBlock1" Text="Hello, world!"
                   local:AttachedProps.MyStr="This is my string!!!"
                   FontSize="30" 
                   Margin="0,10"/>
        <TextBlock x:Name="TextBlock2"
                   local:AttachedProps.MyElement="{Binding ElementName=TextBlock1}"
                   Text="{Binding Path=(local:AttachedProps.MyElement).(local:AttachedProps.MyStr), RelativeSource={RelativeSource Mode=Self}}"
                   FontSize="30"
                   Margin="0,10"/>
    </StackPanel>
```

> AttachedProps.MyElement is of type FrameworkElement, while AttachedProps.MyStr is of type string
In WPF such binding would work
> Any help will be greatly appreciated

**Comment by jeromelaban** [comment]
It would be best to validate if such a scenario works properly on WinUI first. If the original poster has the ability to provide a repro, it will be ideal. In practice, it should be working, but maybe there's some smaller part that is not working.

**Reply by c-npolyak** [reply]
I'll provide a repro

**Reply by c-npolyak** [reply]
please, find the reproduction at https://github.com/c-npolyak/UnoSamples/tree/main/ChainedAttachedPropsIssue. 
The first TextBlock (foreground is Red) does not show because of the chained Attached Properties
The subsequent TextBlocks with Foreground Blue and Black - do show because their binding Paths consist of simple property followed by an attached property and vice versa. 

BTW the sample is WinUI/Uno

**Reply by jeromelaban** [reply]
Thanks, just to be sure, does your sample do what you expect with the `.Windows` head?

**Reply by c-npolyak** [reply]
What is it expected to do?

**Reply by c-npolyak** [reply]
It runs the application and it shows two TextBlocks out of 3

**Reply by c-npolyak** [reply]
I created it as a Uno app.

**Reply by jeromelaban** [reply]
You tell me :) Is it doing what you expect? If not, it's a feature that Uno is unlikely to support (if it's not an authoring error) and there's a need to find a different resolution.

**Reply by c-npolyak** [reply]
I do not understand - you wanted me to reproduce a problem right, I reproduced it. The failing binding works fine in WPF and Avalonia

**Reply by jeromelaban** [reply]
As Uno is a cross-platform implementation of Windows's WinAppSDK, the idea behind reproducing issues is about telling if Uno is not reproducing the original WinUI behavior for a feature. 

If `.Windows` does not do what you expect, it's a missing WinUI feature and Uno will likely not do it as well. If `.Windows` does what you expect, this is an Uno issue that may be able to fix.

Depending on what you find on the `.Windows` head, the resolution is different.

**Reply by c-npolyak** [reply]
My guess is that in UWP it will not work either, though I never tried. Very few things work in UWP :) But I thought Uno provides its own implementation of the bindings and a lot of other stuff too is re-implemented. Including things that are multiplatform (attached properties and bindings are all multiplatform concepts and implementations).

**Reply by jeromelaban** [reply]
I took a quick look and indeed, the first binding expression is not supported by WinUI: 
```xaml
<TextBlock Text="{Binding Path=(local:AttachedProps.MyElement).(local:AttachedProps.MyString), ElementName=TheStackPanel}"
           Margin="20"
           FontSize="30" 
           Foreground="Black"/>
```
and it fails with the following error:
```xaml
Error: BindingExpression path error: '(local:AttachedProps' property not found on 'Microsoft.UI.Xaml.Controls.Grid'. 
BindingExpression: Path='(local:AttachedProps.MyElement).(local:AttachedProps.MyString)' DataItem='Microsoft.UI.Xaml.Controls.StackPanel'; 
target element is 'Microsoft.UI.Xaml.Controls.TextBlock' (Name='null'); target property is 'Text' (type 'String')
```

Yet, you can use `x:Bind` instead:
```xaml
<TextBlock Text="{x:Bind TheStackPanel.(test:AttachedProps.MyElement).(test:AttachedProps.MyString), Mode=OneWay}"
           Margin="20"
           FontSize="30" 
           Foreground="Black"/>
```
Though this one does not work with Uno and that is something we can fix.

**Reply by jeromelaban** [reply]
Also, are you interested in running using the `.Windows` head or are you only interested in the other heads?

**Reply by c-npolyak** [reply]
Yes, I am running both Windows and Wasm

**Reply by c-npolyak** [reply]
Concerning Bind, I always thought it was less powerful than Binding. Will Bind also work with TemplatedParent?

**Reply by jeromelaban** [reply]
It depends on the scenario, what scenario are you looking at, specifically? Also, x:Bind is not less powerful, it's actually more powerful in some areas, particularly regarding [function bindings](https://platform.uno/docs/articles/features/windows-ui-xaml-xbind.html?q=bind). You can use either Binding or x:Bind, depending on the scenario.

**Reply by c-npolyak** [reply]
I'll try to learn more about Bind and please, add the Bind functionality for chaining attached properties to Uno.
